### ISSUE TITLE
Consider allowing SITE_ROOT to be used to configure "base path" as well

### ISSUE THREAD

**Issue by chrishoage** [healthchecks/healthchecks#1091]
I would like to host healthchecks at `example.com/healthchecks` 

I initially thought `SITE_ROOT` would allow for this, however in reality `SITE_ROOT` is more like `SITE_ORIGIN` 

Please consider allowing for `SITE_ROOT=https://example.com/healthchecks` which would configure the routes to use `/heathchecks` (or whatever the user configured). Alternatively a new config option could be added: `BASE_PATH` where it can default to `/` and be configured to the user. In both cases this config would likely need to be communicated to the UI in order to construct links and properly load assets. 

If this request is out of scope for this project then please consider clarifying in the documentation that `SITE_ROOT` can only be used for configuring the origin of the URL, not the full root of the hosted instance as it was not initially clear I couldn't configure the full "root" I wanted using this config option.

**Comment by cuu508** [comment]
Thanks for the suggestion @chrishoage! I updated docs to say `SITE_ROOT` is only used for formatting absolute URLs, and serving on a subpath is not tested and not documented.

**Comment by cuu508** [comment]
I've been experimenting with serving Healthchecks on a subpath the whole day 😵‍💫

There are quite a few moving parts that need to be aligned for things to work:

* the reverse proxy must pass `SCRIPT_NAME` to uwsgi/gunicorn
* uwsgi/gunicorn must chop `SCRIPT_NAME` off from `PATH_INFO`. gunicorn does this by default, uwsgi can be configured to do it:

```
# Read SCRIPT_NAME from HTTP request header "SCRIPT_NAME"
route-run = setscriptname:${HTTP_SCRIPT_NAME}
# If PATH_INFO starts with SCRIPT_NAME, chop it off
route-run = fixpathinfo:
```

* On Healthchecks side, STATIC_URL must contain the subdirectory (or else the paths to static files will be wrong)
* On Healthchecks side, FORCE_SCRIPT_NAME must be set or else management commands like `sendalerts` will generate incorrect URLs

@chrishoage are you using a docker image, or running Healthchecks the regular way, with out containers? I don't think I can provide instructions for ad-hoc setups, but perhaps I can put something together for the docker image, as there I at least control the uwsgi configuration. 

You intuitively expected that you could add a subpath to SITE_ROOT and things would just work. I think we could try and aim for that – the site operator sets SITE_ROOT, and the docker image figures out all the rest of needed settings based off `SITE_ROOT`. No guarantees that I will get this to work, but worth a shot.

**Comment by cuu508** [comment]
cc @s1shed, @netopsengineer, @immanuelfodor.

**Comment by chrishoage** [comment]
Thank you! 

I was unaware of `SCRIPT_NAME` / `FORCE_SCRIPT_NAME` (I'm just generally unfamiliar with django). I'll give that a shot! 

I received some more comments in my email that I don't see here, I presume they were deleted. 

I am running healthchecks in a docker container though a caddy proxy. I protect my "admin" services (of which I consider healthchecks one of) though Authelia under a special sub domain, which is why I was trying to see if SITE_ROOT would let me specify a sub path. 

Much appreciated for the documentation update, and thank you for having healthchecks be selfhost-able!

**Comment by chrishoage** [comment]
To any future google spelunkers who may come across this - it is possible to make this work! 

## Steps: 

1. In local_settings.py add the following 

```python
FORCE_SCRIPT_NAME = "/checks/"
STATIC_URL = "/checks/static/
PING_ENDPOINT = "https://example.com/checks/ping/
```

2. In your proxy it is important to configure the proxy to strip /checks. Example Caddy config:

```Caddyfile
example.com {
  rewrite /checks /checks/
  route /checks/* {
    uri strip_prefix /checks
    reverse_proxy healthchecks:8000
  }
}
```

That should be all that is required to make this _function_. 

## Notes 

I am using REMOTE_USER_HEADER and Authelia - I have no idea if the above methods allow login to work. YMMV 

In order for this to be "officially supported" (ideally users wouldn't need to strip the path before feeding it downstream) by a `BASE_PATH` config in `local_settings.py` the following would need to happen

1. Auto configure `FORCE_SCRIPT_NAME`, `STATIC_URL` and `PING_ENDPOINT` based on a `BASE_PATH` value. This could likely default to `/` to make life easy 
2. Update [`hc/urls.py` ](https://github.com/healthchecks/healthchecks/blob/master/hc/urls.py) to "re-parent" the routes. `path(BASE_PATH, urlpatterns),` perhaps, that's just a guess (never worked with Django) 

If someone enterprising enough wishes to open a PR it sounds like this project may accept it. However since I was able to make this function enough for my own use, I'm happy with the solution I landed at above. 

Huge thanks to @cuu508 for the nudges that led me to be able to google what I needed to know to make the above work.

**Comment by cuu508** [comment]
> I received some more comments in my email that I don't see here, I presume they were deleted.

Yes, I was (and still am) going forth and back between whether I want to spend effort on this or not.

**Comment by chrishoage** [comment]
@cuu508 following your example (adapted for my Caddy and Docker Compose setup) I am able to get a sub path to work by simply supplying `SITE_ROOT`! 

For you to see, here is what I have

```yaml
  healthchecks:
    container_name: healthchecks
    hostname: healthchecks
    image: healthchecks/healthchecks:master
    restart: ${RESTART-unless-stopped}
    volumes: # I left off a /data volume since for me this was an ephemeral test 
      - /etc/localtime:/etc/localtime:ro
    environment:
      - DB=sqlite
      - DB_NAME=/data/hc.sqlite
      - DEBUG=False
      - SITE_ROOT=https://example.com/checks
      - ALLOWED_HOSTS=healthchecks,example.com
      - EMAIL_HOST=smtp.sendgrid.net
      - EMAIL_HOST_USER=apikey
      - EMAIL_HOST_PASSWORD=${SG_MAIL_KEY:?required}
      - DEFAULT_FROM_EMAIL=noreply@example.com
      - REMOTE_USER_HEADER=HTTP_REMOTE_EMAIL
      - SECRET_KEY=askldfjaslkdfjasdlkfjasdklasdlkfjlska # example key 
```

```Caddyfile
# simplified for this demonstration  
example.com {
  rewrite /checks /checks/
  reverse_proxy /checks/ healthchecks:8000
}
```

In my testing, this works great! (honestly the workaround you pointed me to above was also functioning, but I'll never say no to first class support for the feature!) 

Please let me know if I can help any further!

**Comment by cuu508** [comment]
I've published v3.8. It has the support for path in SITE_ROOT, and has a few additional fixes related to serving Healthchecks on a path:

* The URLs in email notifications  were missing the path
* Redirect to login page was missing the path and so redirecting the user to a 404 page (originally reported in #382)
* The redirect *after* login was not working correctly too

I've now also removed the temporary image tagged as "master" from Docker Hub.

**Comment by chrishoage** [comment]
I went to update today to pull in these changes but I am no longer able to make it work. 

I end up with the application thinking I'm loading `http://healthchecks:8000/checks/checks` when curling `http://healthchecks:8000/checks`

example (I am using the linuxserver docker image, but when I started having problems I switched back to the one distributed from this repo, and I have the same behavior): 

```
❯ sudo docker exec -it healthchecks /bin/bash
root@healthchecks:/# curl http://healthchecks:8000/checks
<!DOCTYPE html>
<html lang="en">
<head>
  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  <title>Page not found at /checks</title>
  <meta name="robots" content="NONE,NOARCHIVE">
  <style>
    html * { padding:0; margin:0; }
    body * { padding:10px 20px; }
    body * * { padding:0; }
    body { font-family: sans-serif; background:#eee; color:#000; }
    body > :where(header, main, footer) { border-bottom:1px solid #ddd; }
    h1 { font-weight:normal; margin-bottom:.4em; }
    h1 small { font-size:60%; color:#666; font-weight:normal; }
    table { border:none; border-collapse: collapse; width:100%; }
    td, th { vertical-align:top; padding:2px 3px; }
    th { width:12em; text-align:right; color:#666; padding-right:.5em; }
    #info { background:#f6f6f6; }
    #info ol { margin: 0.5em 4em; }
    #info ol li { font-family: monospace; }
    #summary { background: #ffc; }
    #explanation { background:#eee; border-bottom: 0px none; }
    pre.exception_value { font-family: sans-serif; color: #575757; font-size: 1.5em; margin: 10px 0 10px 0; }
  </style>
</head>
<body>
  <header id="summary">
    <h1>Page not found <small>(404)</small></h1>

    <table class="meta">
      <tr>
        <th scope="row">Request Method:</th>
        <td>GET</td>
      </tr>
      <tr>
        <th scope="row">Request URL:</th>
        <td>http://healthchecks:8000/checks/checks</td>
      </tr>
```

Strangely however, the actual debug logs show 

```
[pid: 159|app: 0|req: 6/6] 172.18.0.28 () {60 vars in 1084 bytes} [Mon Dec 16 02:35:11 2024] GET /checks/ => generated 10887 bytes in 11 msecs (HTTP/1.1 404) 7 headers in 230 bytes (1 switches on core 0)
```

For now I've gone back to the solution outlined in https://github.com/healthchecks/healthchecks/issues/1091#issuecomment-2516087227 which is working for me. However I think the last round of changes after I left my message may have broken something, as I am no longer able to make it work with the config outlined in https://github.com/healthchecks/healthchecks/issues/1091#issuecomment-2518991756

**Comment by cuu508** [comment]
@chrishoage did you pass `SITE_ROOT` as an environment variable to the container?

Specifying it in local_settings.py does not work, because [uWSGI also needs to access it](https://github.com/healthchecks/healthchecks/blob/14bcc84a781c91584e17ff9d1944465169fcee9a/docker/uwsgi.ini#L39). 

Here's how I intended it to work:

* client requests http://healthchecks:8000/checks
* uWSGI sets SCRIPT_NAME to "/checks"
* uWSGI also sees PATH_INFO starts with "/checks" and chops it off
* python code sees PATH_INFO="", SCRIPT_NAME="/checks"

The linuxserver image doesn't have the SCRIPT_NAME and PATH_INFO patching bits in their uwsgi.ini so I would not expect it to work.

**Comment by chrishoage** [comment]
I tried to reproduce the behavior I was seeing on a different host and it was indeed working. 

I think my issue was I had an outdated image on my production host. While at the time I thought I had pulled I obviously didn't, so must have been running an outdated image of `healthchecks/healthchecks:latest`

After pulling `healthchecks/healthchecks:latest` I am no longer able to reproduce. I will use this image from now on - thank you! 

I apologize for the trouble!
